# Elegans/Briggsae Embryo Single-Cell Dataset
Christopher Large, Rupa Khanal, LaDeana Hillier, Junhyong Kim, John I. Murray, and Robert Waterston

In anticipation of our upcoming manuscript describing a comparitive analysis of single-cells from _C. elegnas_ and _C. briggsae_ embryogenesis, we are releasing multiple forms of data summaries that we hope will be useful to the community.

__Preprint: [Link upcoming]__

The avaliable data are:
 - A VisCello Object
 - Cell-type summaries (Both visual descriptions and tables)
 - Gene summaries (Both visual descriptions and tables)
 - Gene expression (TPM) tables
 - Cell-marker tables

### VisCello Object:
VisCello is a single-cell gene expression visualization tool generated by Qin Zhu during his doctoral work at the University of Pennsylvania. The software required to load these data into VisCello is avaliable for R at this GitHub: https://github.com/qinzhu/VisCello

After installing VisCello, you can load the library and point it to your local downloaded copy of this repo. Please note that the data size is immense and will require some patience while loading.

#### How to install and load the data into VisCello:

``` r
install.packages("devtools") 
devtools::install_github("qinzhu/VisCello") # install
library(VisCello) # load
cello("/Path/To/The/Downloaded/Data/Cello/") # launch with elegans briggsae data
```
<details>
           <summary>VisCello Example</summary>
           <p>Content 1 Content 1 Content 1 Content 1 Content 1</p>
</details>

### Cell Summary Info:
 1. The relative TPM of every gene in elegans and briggsae
 2. A barplot of the cell type markers for that cell type using the WormCat gene categories as a descriptor
 3. The top cell type markers that are shared (black outline), private to elegans (green), or private to briggsae (blue)
 4. A bunch of cell type metrics, where the values for that cell type are shown in green for elegans and blue for briggsae (red for both) ontop of the dataset wide distribution
  - Cell count: number of cells in the dataset
  - Gini coeficient: A measure of inequality that shows how evenly distributed the TPM values are (0 = even, 1 = skewed)
  - The number of genes ‘detected’ in that cell type. Calculated by generating 1000 bootstraps of the TPM, then selecting genes whose 95% lower CI doesn’t intersect 0
  - How many of the markers of that cell type are just in one species versus the total markers (shared + private)
  - Number of UMI’s of that cell type
  - Jensen-Shannon Distance: Metric of distance between the two species cell transcriptomes
  - Pearson Correlation: Metric of similarity between the two species cell transcriptomes
  - The number of differentially expressed genes between the species


<details>
           <summary>Cell Plot Example</summary>
           <p>Content 1 Content 1 Content 1 Content 1 Content 1</p>
</details>

### Gene Summary Info:
 1. The relative TPM of every gene in elegans and briggsae
 2. A barplot of the cell type markers for that cell type using the WormCat gene categories as a descriptor
 3. The top cell type markers that are shared (black outline), private to elegans (green), or private to briggsae (blue)
 4. A bunch of cell type metrics, where the values for that cell type are shown in green for elegans and blue for briggsae (red for both) ontop of the dataset wide distribution
  - Cell count: number of cells in the dataset
  - Gini coeficient: A measure of inequality that shows how evenly distributed the TPM values are (0 = even, 1 = skewed)

### TPM tables:

### Cell-marker tables:
