# C. elegans/C. briggsae Embryo Single-Cell Dataset
Christopher R. L. Large, Rupa Khanal, LaDeana Hillier, Junhyong Kim, John I. Murray, and Robert Waterston

In anticipation of our upcoming manuscript describing a comparitive analysis of single-cells from _C. elegans_ and _C. briggsae_ embryogenesis, we are releasing multiple forms of data summaries that we hope will be useful to the community.

__Preprint: [Link upcoming]__

The avaliable data are:
 - A VisCello Object
 - Cell-type summaries (Both visual descriptions and tables)
 - Gene summaries (Both visual descriptions and tables)
 - Gene expression (TPM) tables
 - Cell-marker tables

### How to download:
Due to large data file currently being hosted on Git LFS, you cannot use git clone to download these files. Please follow the instructions below below to download the dataset in its entirety:

 1. Install git from:
 * Windows: <https://gitforwindows.org/>
 * MacOS: <https://git-scm.com/download/mac>
 * Linux: <https://git-scm.com/download/linux>

 2. Install git-lfs from:
 * Windows: Git LFS is included in the distribution of Git for Windows.
 * MacOS and Linux: <https://git-lfs.github.com/>

 It is also possible to install git-lfs using brew or conda with:
```
brew install git-lfs
git lfs install
```
```
conda install -c conda-forge git-lfs
git lfs install
```

 3. Clone the github repository using this command in terminal:
    
```
git lfs clone https://github.com/livinlrg/C.elegans_C.briggsae_Embryo_Single_Cell.git
```

### VisCello Object:
VisCello is a single-cell gene expression visualization tool generated by Qin Zhu during his doctoral work at the University of Pennsylvania. The software required to load these data into VisCello is avaliable for R at this GitHub: https://github.com/qinzhu/VisCello

After installing VisCello, you can load the library and point it to your local, downloaded copy of this repository. Please note that the data size is large and will require some patience while loading.

#### How to install and load the data into VisCello:

``` r
install.packages("devtools") 
devtools::install_github("qinzhu/VisCello") # install
library(VisCello) # load
cello("/Path/To/The/Downloaded/Data/Cello/") # launch with elegans briggsae data
```


<details>
           <summary>VisCello Information</summary>
<p>

<p align="center">
  <img width="800" src=/Example_Images/VisCelloExample.png>
</p>

 #### Summary:
<p>
 Using VisCello, you can visualize the expression of your favorite gene across the <em>C. elegans</em> and <em>C. briggsae</em> single-cell datasets in a number of different UMAP projections. Here we have included joint projections of the global dataset, cell class subsets, and time subsets. All of the joint projections are generated using Seurat V4 CCA. Additionally, we have included projections in which the <em>C. elegans</em> and <em>C. briggsae</em> cells are on their own in the same space as the joint projections.
</p>

<p>
 In addition to viewing expression, you can visualize the cell identity and other metadata that we have annotated using orthologous markers between <em>C. elegans</em> and <em>C. briggsae</em>. The information about these metadata columns is below.
</p>
 

#### Meta Data:
- lineage - Manually annotated cellular lineage. For ambiguities in division orientation, an x is used (e.g. MSx to refer to MSa and MSp).
- cell_type - Manually annotated terminal cell type identity. 
- species - Whether the cell is from <em>C. elegans</em> or <em>C. briggsae</em>
- embryo_time - The estimated age of the embryo from which the cell was drawn. See Packer and Qin et al., 2019 for more details on how this was caluclated. <em>C. briggsae</em> embryo_time was estimated using the orthologous genes between the species.
- dataset - Which collection batch the cells come from. 
- n_umi - 
- genotype
- potential_low_quality_cell
- high_background
- possible_doublet
- packer_cell_type
- packer_cell_subtype
- packer_plot_cell_type
- SizeFactor
- smoothed_embryo_time
- embryo_time_bin
- Gene Expression
</p>
</details>


### Cell Summary Info:
 1. The relative TPM of every gene in elegans and briggsae
 2. A barplot of the cell type markers for that cell type using the WormCat gene categories as a descriptor
 3. The top cell type markers that are shared (black outline), private to elegans (green), or private to briggsae (blue)
 4. A bunch of cell type metrics, where the values for that cell type are shown in green for elegans and blue for briggsae (red for both) ontop of the dataset wide distribution
  - Cell count: number of cells in the dataset
  - Gini coeficient: A measure of inequality that shows how evenly distributed the TPM values are (0 = even, 1 = skewed)
  - The number of genes ‘detected’ in that cell type. Calculated by generating 1000 bootstraps of the TPM, then selecting genes whose 95% lower CI doesn’t intersect 0
  - How many of the markers of that cell type are just in one species versus the total markers (shared + private)
  - Number of UMI’s of that cell type
  - Jensen-Shannon Distance: Metric of distance between the two species cell transcriptomes
  - Pearson Correlation: Metric of similarity between the two species cell transcriptomes
  - The number of differentially expressed genes between the species


<details>
           <summary>Cell Plot Example</summary>
           <p>Content 1 Content 1 Content 1 Content 1 Content 1</p>
</details>

### Gene Summary Info:
 1. The relative TPM of every gene in elegans and briggsae
 2. A barplot of the cell type markers for that cell type using the WormCat gene categories as a descriptor
 3. The top cell type markers that are shared (black outline), private to elegans (green), or private to briggsae (blue)
 4. A bunch of cell type metrics, where the values for that cell type are shown in green for elegans and blue for briggsae (red for both) ontop of the dataset wide distribution
  - Cell count: number of cells in the dataset
  - Gini coeficient: A measure of inequality that shows how evenly distributed the TPM values are (0 = even, 1 = skewed)

### TPM tables:

### Cell-marker tables:
