# _C. elegans_ and _C. briggsae_ Embryo Single-Cell Dataset
Christopher R. L. Large, Rupa Khanal, LaDeana Hillier, Chau Huynh, Junhyong Kim, John I. Murray, and Robert Waterston

In anticipation of our upcoming manuscript describing a comparative analysis of homologous cell-types between _C. elegans_ and _C. briggsae_ across embryogenesis using single-cell RNA-sequencing, we are releasing an accessible collection of data summaries on the homologous genes and cell-types between these species.

__Preprint: [Link upcoming]__


The avaliable data are as follows. A detailed description of these data is below in dropdown menus.
 - A VisCello Object
 - Cell-type summaries (Both visual descriptions and tables)
 - Gene summaries (Both visual descriptions and tables)
 - Gene expression (TPM) tables
 - Cell-marker tables


### How to download:
Due to how large the data files are, they are currently being hosted on Git LFS (Large File Service). This means you cannot use the default git clone to download these files. Instead, you will need to use git LFS. If the use of terminal commands proves too inaccessible, it is also feasible to download individual files of interest using githubs web browser. Please follow the instructions below below to download the dataset in its entirety:

 1. Install git from:
 * Windows: <https://gitforwindows.org/>
 * MacOS: <https://git-scm.com/download/mac>
 * Linux: <https://git-scm.com/download/linux>

 2. Install git-lfs from:
 * Windows: Git LFS is included in the distribution of Git for Windows.
 * MacOS and Linux: <https://git-lfs.github.com/>

 It is also possible to install git-lfs using brew or conda with:
```
brew install git-lfs
git lfs install
```
```
conda install -c conda-forge git-lfs
git lfs install
```

 3. Clone the github repository using this command in terminal:
    
```
git lfs clone https://github.com/livinlrg/C.elegans_C.briggsae_Embryo_Single_Cell.git
```

### VisCello Object:
VisCello is a single-cell gene expression visualization tool generated by Qin Zhu during his doctoral work at the University of Pennsylvania. The software required to load these data into VisCello is avaliable for R at this GitHub: https://github.com/qinzhu/VisCello

After installing VisCello, you can load the library and point it to your local, downloaded copy of this repository. Please note that the data size is large and will require some patience while loading.

#### How to install and load the data into VisCello in R:

``` r
# Inside of an R session, run the follow commands
install.packages("devtools") 
devtools::install_github("qinzhu/VisCello") # install
library(VisCello) # load
cello("/Path/To/The/Downloaded/Data/Cello/") # launch with elegans briggsae data
```

<details>
           <summary>VisCello Information</summary>
<p>

<p align="center">
  <img width="800" src=/Example_Images/VisCelloExample.png>
</p>

 #### Summary:
<p>
 Using VisCello, you can visualize the expression of your favorite gene across the <em>C. elegans</em> and <em>C. briggsae</em> single-cell datasets in a number of different UMAP projections. Here we have included joint projections of the global dataset, of individual cell class subsets, and of embryo time subsets. All of the joint projections are generated using Seurat V4 CCA. We have also included projections in which the <em>C. elegans</em> and <em>C. briggsae</em> cells are displayed on their own in the same UMAP embeddings.
</p>

<p>
You can search for your favorite gene from either of the species by looking up the gene name, the sequence name, or the WBGene name.
</p>

<p>
 In addition to viewing expression, you can also visualize the cell identity and other metadata that we have annotated using the orthologous markers between <em>C. elegans</em> and <em>C. briggsae</em>. The information about these metadata columns is below.
</p>
 
#### Meta Data:
- lineage - The manually annotated cell lineage. For ambiguities in division orientation, an x is used (e.g. MSx to refer to MSa and MSp).
- cell_type - The terminal cell-type identities, manually annotated using homologous marker genes. 
- species - Whether the cell is from <em>C. elegans</em> or <em>C. briggsae</em>
- embryo_time - The estimated age of the embryo from which the cell was drawn. See Packer and Zhu et al., 2019 for more details on how this was caluclated. <em>C. briggsae</em> embryo_time was estimated using the orthologous genes between the species.
- dataset - Which collection batch the cells come from. 
- n_umi - The number of UMI-collapsed sequencing reads that are associated with the cell.
- genotype - The genotype from which the cell came from. Some of the _C. elegans_ cells are from mutant animals.
  - Wild-type _C. elegans_: N2 and VC2010
  - Wild-type _C. briggsae_: AF16
  - Mutant _C. elegans_ for mec-3: VC2396 mec-3(gk1126). Mutants for mec-3 appear to be missing their touch neurons and markers of the touch neurons are not detected.
  - Mutant _C. elegans_ for M03D4.4: VC4183 M03D4.4(gk5269[loxP + myo-2p::GFP::unc-54 3' UTR + rps-27p::neoR::unc-54 3' UTR + loxP]). This mutant strain for M03D4.4 appears otherwise wild-type in cell composition and expression.
  - Mutant _C. elegans_ for ceh-9: YL633 ceh-9(tm2747). This mutant strain for ceh-9 appears otherwise wild-type in cell composition and expression.
- potential_low_quality_cell - Using a variety of manual annotation stratagies, we have identified some cells that don't behave consistently across UMAP embeddings due to a variety of techincal reasons. These have been left in the dataset as they often represent 'normal cells', but have been labeled as being potentially low-quality.
- high_background - The amount of background reads was estimated for every cell similar to Packer and Zhu et al., 2019. The cells labeled here as TRUE had a fraction of reads from background higher than 0.75.
- possible_doublet - Droplets that annotated as possibly containing two or more cells. Not all cells annotated as possible droplets are as such. Please see Packer and Zhu et al., 2019 for details on how the background was estimated.
- packer_cell_type - Cell type annotation from Packer and Zhu et al., 2019.
- packer_cell_subtype - Cell type annotation from Packer and Zhu et al., 2019.
- packer_plot_cell_type - Cell type annotation from Packer and Zhu et al., 2019.
- SizeFactor - A column used to estimate the library size.
- smoothed_embryo_time - The estimated embryo time calculated as above, with an additional nearest neighbor smoothing algorithm to use the neighobring cell's embryo time and transcriptome to better approximate the age of the embryo.
- embryo_time_bin - Binned smoothed embryo time with lt_100 meaning 'less than 100' and gt_710 meaning 'greater than 710.
- Gene Expression - Used to inspect gene expression.
---
</p>
</details>

<details>
           <summary>Cell Summary Information</summary>
<p>
 <p>

 The cell-type summaries avaliable in the cell_plots/ directory describe how the cell-types differ between _C. elegans_ and _C. briggsae_ using a variety of summary statistics. The description of what is included in these summaries are below with an example plot for the ASG neuron:

 <p align="center">
  <img width="800" src=/Example_Images/ExampleASGNeuron.png>
</p>
 
 1. The relative TPM of every gene in _C. elegans_ and _C. briggsae_. Whether a gene is a cell-type marker within that species or both is labeled.
 2. A barplot of the cell-type markers from _C. elegans_ binned by their WormCat gene category.
 3. The top cell-type markers that are shared between species (black outline), private to _C. elegans_ (green), or private to _C. briggsae_ (blue). The private markers can also include genes that weren't annotated as being directly orthologous between the species.
 4. A bunch of cell type metrics, where the values for that cell type are shown in green for _C. elegans_ and blue for _C. briggsae_ (red for both) ontop of the dataset wide distribution.
  - Cell count: number of cells in the dataset
  - Gini coeficient: A measure of inequality that shows how evenly distributed the TPM values are (0 = even, 1 = skewed)
  - The number of genes ‘detected’ in that cell type. Calculated by generating 1000 bootstraps of the TPM, then selecting genes whose 95% lower CI doesn’t intersect 0.
  - How many of the markers of that cell-type are just in one species versus the total markers (shared + private)
  - The number of UMI-collapsed sequencing reads that are associated with the cell-type.
  - Jensen-Shannon Distance: Metric of distance between the two species cell transcriptomes.
  - Pearson Correlation: Metric of similarity between the two species cell transcriptomes.
  - The number of differentially expressed genes between the species.
---
</p>
</p>
</details>

</p>
</details>

<details>
           <summary>Gene Summary Information: </summary>
<p>
 <p>
The gene summaries that will become avaliable in the gene_plots/ directory describe how the gene expression profiles differ between _C. elegans_ and _C. briggsae_ using a variety of summary statistics. The description of what is included in these summaries are below with an example plot for pha-4:

 1. Global UMAP showing the expression of your gene of interest. 
 2. Cell subset UMAP showing the expression of your gene of interest. The choice of which UMAP is shown is based on which cell-type shows maximum expression across all cell-types between the two species.
 3. Terminal cell-type comparative TPM values shown in log2 space. The cell-types are summarized by their cell class.
 4. Progenitor cell-type comparative TPM values shown in log2 space. The progenitors are summarized by their general lineage type.
 5. A bunch of gene metrics, where the values for this gene is shown in red as a confidence interval (CI) range on top of the dataset wide distribution. These metrics are shown for the terminal and progenitor cell-types.
  - Gene expression patern distance shown as the Jensen-Shannon Distance (JSD) calculated on the bootstrapped TPM values. The CI and median for the gene JSD was calculated on the bootstrap resampled TPM values.
  - Gene expression patern distance shown as the Pearson correlation coefficient calculated on the bootstrapped TPM values. The CI and median for the gene Pearson correlation coefficient was calculated on the bootstrap resampled TPM values.
  - The broadness of gene expression pattern shown as the Tau value for _C. elegans_.
  - The broadness of gene expression pattern shown as the Tau value for _C. briggsae_.
  - The maximum TPM value across any cell-type in both species.
 
---
</p>
</p>
</details>


<details>
           <summary>TPM Tables Information: </summary>
<p>
 <p>

Avaliable here is the expression value of every gene in the _C. elegans_ and _C. briggsae_ genome, summarized as the transcripts per million (TPM) on pseudobulked progenitor and terminal cell-types. To evaluate the variation in the measurement of these TPM values, we have used bootstrapping to take samples of the cells and generate several confidence intervals on the cellular expression values. A gene can be thought of as confidently detected in that cell-type if its lower 95% confidence interval does not intersect zero. Additionally, we have assessed the percentage of cells from that cell-type that we were able to detect expression from.
 
---
</p>
</p>
</details>

<details>
           <summary>Differentially expressed-genes between species: </summary>
<p>
 <p>
  
To identify genes differentially expressed between _C. elegans_ and _C. briggsae_ within the homologous cell-types, we used Seurat V4. A Wilcoxon Rank Sum test was run between the cells of that cell-type from _C. elegans_ against the cells of that cell-type from _C. briggsae_. The data have been filtered for an adjusted p-value less than 0.05 and a log2 fold-change of greater than 1 or less than -1. The columns in the table are as below:

 - p_val - The unadjusted p-value between _C. elegans_ cells and _C. briggsae_ cells for that gene in that cell-type.
 - avg_log2FC_elegans_to_briggsae - The log2 fold-change between _C. elegans_ cells and _C. briggsae_ cells for that gene in that cell-type.
 - elegans_cell_fraction - The fraction of cells for which this gene was detected in _C. elegans_.
 - briggsae_cell_fraction - The fraction of cells for which this gene was detected in _C. briggsae_.
 - p_val_adj- The adjusted p-value between _C. elegans_ cells and _C. briggsae_ cells for that gene in that cell-type.
 - gene - The gene that was tested for its differential expression.
 - cell_type - The cell-type in which the gene was tested for its differential expression.
 - elegansTPM - The TPM of that gene in that cell-type.
 - elegansMaxTPM - The maximum TPM of that gene in any cell-type.
 - elegansMeanTPM - The average TPM of that gene across the cell-types.
 - elegansNumberCellTypesDetectedIn_0.05 - The number of cell-types that gene was confidently detected in using a trinirization method that tests whether a gene is confidently detected in at least 5% of the cells of that cell-type.
 - elegansNumberCellTypesDetectedIn_0.10 - TThe number of cell-types that gene was confidently detected in using a trinirization method that tests whether a gene is confidently detected in at least 10% of the cells of that cell-type.
 - elegansNumberCellTypesDetectedIn_Boot - The number of cell-type that gene was confidently, calculated by generating 1000 bootstraps of the TPM, then selecting genes whose 95% lower CI doesn’t intersect 0.
 - briggsaeTPM - The TPM of that gene in that cell-type.
 - briggsaeMaxTPM - The maximum TPM of that gene in any cell-type.
 - briggsaeMeanTPM - The average TPM of that gene across the cell-types.
 - briggsaeNumberCellTypesDetectedIn_0.05 - The number of cell-types that gene was confidently detected in using a trinirization method that tests whether a gene is confidently detected in at least 5% of the cells of that cell-type.
 - briggsaeNumberCellTypesDetectedIn_0.10 - The number of cell-types that gene was confidently detected in using a trinirization method that tests whether a gene is confidently detected in at least 10% of the cells of that cell-type.
 - briggsaeNumberCellTypesDetectedIn_Boot - The number of cell-type that gene was confidently, calculated by generating 1000 bootstraps of the TPM, then selecting genes whose 95% lower CI doesn’t intersect 0.
 - elegansTau - The broadness of the gene expression pattern.
 - briggsaeTau - The broadness of the gene expression pattern.

---
</p>
</p>
</details>

<details>
           <summary>Cell-type marker table information: </summary>
<p>
 <p>
  
To identify genes that are specific to certain cell-types in _C. elegans_ and _C. briggsae_, marker genes were called in both species using Seurat V4. In practice a Wilcoxon Rank Sum test is run between all of the cells of that cell-type against the rest of the dataset to find genes that are significantly enriched in the cell-type. The data have been filtered for an adjusted p-value less than 0.05 and a log2 fold-change of greater than 1 between the cell-type and the dataset. This was calculated for both of the species in the dataset on the terminal cell-types. The columns in the table are as below:

 - p_val.species - The unadjusted p-value between the cells of that cell-type and the cells of rest of the dataset for that species.
 - avg_log2FC.species - The log2 fold-change between the cells of that cell-type and the cells of rest of the dataset for that species.
 - pct.1.species - The fraction of cells for which this gene was detected in that cell-type for that species.
 - pct.2.species - The fraction of cells for which this gene was detected in the rest of the cells in the dataset for that species.
 - p_val_adj.species - The adjusted p-value between the cells of that cell-type and the cells of rest of the dataset for that species.
 - cell-type - The cell-type in which the gene was tested for its marker status.
 - gene - The gene that was tested for its marker status.
 - p_val.other_species - If the gene was tested in the other species for its marker status, the values for that test are shown here.
 - avg_log2FC.other_species - If the gene was tested in the other species for its marker status, the values for that test are shown here.
 - pct.1.other_species - If the gene was tested in the other species for its marker status, the values for that test are shown here.
 - pct.2.other_species - If the gene was tested in the other species for its marker status, the values for that test are shown here.
 - p_val_adj.other_species - If the gene was tested in the other species for its marker status, the values for that test are shown here.
 - elegansTPM - The TPM of that gene in that cell-type.
 - elegansMaxTPM - The maximum TPM of that gene in any cell-type.
 - elegansMeanTPM - The average TPM of that gene across the cell-types.
 - elegansNumberCellTypesDetectedIn_0.05 - The number of cell-types that gene was confidently detected in using a trinirization method that tests whether a gene is confidently detected in at least 5% of the cells of that cell-type.
 - elegansNumberCellTypesDetectedIn_0.10 - TThe number of cell-types that gene was confidently detected in using a trinirization method that tests whether a gene is confidently detected in at least 10% of the cells of that cell-type.
 - elegansNumberCellTypesDetectedIn_Boot - The number of cell-type that gene was confidently, calculated by generating 1000 bootstraps of the TPM, then selecting genes whose 95% lower CI doesn’t intersect 0.
 - briggsaeTPM - The TPM of that gene in that cell-type.
 - briggsaeMaxTPM - The maximum TPM of that gene in any cell-type.
 - briggsaeMeanTPM - The average TPM of that gene across the cell-types.
 - briggsaeNumberCellTypesDetectedIn_0.05 - The number of cell-types that gene was confidently detected in using a trinirization method that tests whether a gene is confidently detected in at least 5% of the cells of that cell-type.
 - briggsaeNumberCellTypesDetectedIn_0.10 - The number of cell-types that gene was confidently detected in using a trinirization method that tests whether a gene is confidently detected in at least 10% of the cells of that cell-type.
 - briggsaeNumberCellTypesDetectedIn_Boot - The number of cell-type that gene was confidently, calculated by generating 1000 bootstraps of the TPM, then selecting genes whose 95% lower CI doesn’t intersect 0.
 - elegansTau - The broadness of the gene expression pattern.
 - briggsaeTau - The broadness of the gene expression pattern.
 - InOtherSpecies - Whether this gene marker is also a marker in the other species.
 - orthologous_gene - Whether the gene was found to be orthologous between the two species by WormBase and a synteny recovery method.
 - elegansSinglet - Whether the gene was found as a 1:x ortholog in _C. elegans_ by OrthoFinder.
 - briggsaeSinglet - Whether the gene was found as a 1:x ortholog in _C. briggsae by OrthoFinder.
 - orthogroupSize - The size of the orthogroup (number of genes in that 'family').
 - orthogroupName - The name of the orthogroup (for internal use).
 - WormCat.1 - The WormCat category of this gene at a tier one level.
 - WormCat.2 - The WormCat category of this gene at a tier two level.
 - WormCat.3 - The WormCat category of this gene at a tier three level.

---
</p>
</p>
</details>

